extensions/avro/src/main/java/com/hazelcast/jet/avro/impl/AvroSchemaStorage.java

```diff
@@ -0,0 +1,119 @@
+
+/**
+ * Schemas are stored in an IMap, where the key is the 64-bit Rabin fingerprint of the
+ * canonical JSON representation of the schema. As explained in the
+ * <a href="https://avro.apache.org/docs/1.11.1/specification/#schema-fingerprints">Avro
+ * specification</a>, such fingerprints are safe to be used as a key in schema caches of up
+ * to a million entries (for such a cache, the chance of a collision is 3x10<sup>-8</sup>).
+ * <p>
+ * Schemas that are not used for one week are dropped and the expiration time of the most
+ * recently used schemas are updated once an hour in batches.
+ */
+public class AvroSchemaStorage {
+    private static final String AVRO_SCHEMAS_MAP_NAME = INTERNAL_JET_OBJECTS_PREFIX + "avro.schemas";
+    private static final int AVRO_SCHEMAS_MAX_IDLE_SECONDS = (int) DAYS.toSeconds(7);
+
+    private final Map<Long, Schema> localSchemas = new ConcurrentHashMap<>();
+    private final Map<Schema, Long> localSchemaIds = new ConcurrentHashMap<>();
+
+    private final ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();
+    private final AtomicReference<Map<Long, Schema>> mostRecentlyUsed = new AtomicReference<>(new HashMap<>());
+
+    private IMap<Long, Schema> schemas;
+
+    public AvroSchemaStorage(HazelcastInstance instance) {
+        instance.getLifecycleService().addLifecycleListener(event -> {
+            if (event.getState() == LifecycleEvent.LifecycleState.STARTED) {
+                instance.getConfig().addMapConfig(new MapConfig(AVRO_SCHEMAS_MAP_NAME)
+                        .setMaxIdleSeconds(AVRO_SCHEMAS_MAX_IDLE_SECONDS));
+                schemas = instance.getMap(AVRO_SCHEMAS_MAP_NAME);
+                schemas.addEntryListener(new EntryAdapter<Integer, String>() {
+                    @Override
+                    public void onEntryEvent(EntryEvent<Integer, String> e) {
+                        long schemaId = e.getKey();
+                        Schema schema = new Schema.Parser().parse(e.getValue());
+                        switch (e.getEventType()) {
+                            case ADDED:
+                                localSchemas.put(schemaId, schema);
+                                localSchemaIds.put(schema, schemaId);
+                                break;
+                            case EVICTED:
+                                localSchemas.remove(schemaId);
+                                localSchemaIds.remove(schema);
+                                break;
+                            default:
+                                throw new UnsupportedOperationException(
+                                        "Avro Schema Storage does not support " + e.getEventType());
+                        }
+                    }
+                }, true);
+
+                // Update the expiration time of most recently used schemas periodically.
+                executorService.scheduleAtFixedRate(() ->
```

Comment:
IMap has expiration policies, looks like reinventing the wheel, because we cache IMap in local CHM.